### PR TITLE
Linux tv-casting-app v1.3 Commissioner-Generated passcode follow up

### PR DESCRIPTION
**Pull request by pgregorr-amazon**
Linux tv-casting-app example app, addressed post merge comments on PR 33479 and changed the CastingPlayer endpoint for demo interactions.

**Change summary**

1.	In CastingPlayer.h, changed “const unsigned long long int kCommissioningWindowTimeoutSec” to “const uint16_t kCommissioningWindowTimeoutSec” per andy31415 from Google.
2.	In CastingPlayer.h, removed the parameters from ContinueConnecting() since we can use the parameters that were passed to VerifyOrEstablishConnection().
3.	Moved some error checking, verifying the IdentificationDeclarationOptions, in CastingPlayer.cpp which needs to be verified prior to the IsFailSafeArmed() check.
4.	In simple-app-helper.cpp added a gCommissionerGeneratedPasscodeFlow Boolean to control which Endpoint we perform demo interactions on. 
5.	Also, in simple-app-helper.cpp changed the connectedhomeip/examples/tv-app target app Vendor ID for the Commissioner Generated Passcode flow to (1111). The default target app Vendor ID previously used (65521) has Account Login implemented and would commission with that by default without displaying the Commissioner Generated Passcode.
6.     Added a copy constructor and assignment operator to CastingPlayer.h per sharadb-amazon request.

**Testing**

Verified and tested locally with the Linux, Android and iOS tv-casting-app example mobile apps, and the Linux tv-app (CastingPlayer). With the Linux example tv-casting-app app, we are able to successfully commission using the commissioner generated passcode flow as follows:

1.	Start tv-app:                connectedhomeip % ./out/tv-app/chip-tv-app
2.	Start tv-casting-app:   connectedhomeip % ./out/tv-casting-app/chip-tv-casting-app
3.	In tv-casting-app:       > cast request 0 commissioner-generated-passcode
4.	In tv-app:                    > controller ux ok
5.	Tv-app displays the commissioner-generated-passcode in hex: Casting passcode: [0x00BC_614E] or 12345678 in decimal.
7.	In tv-casting-app:       > cast setcommissionerpasscode 12345678
8.	The logs display “simple-app-helper.cpp::ConnectionHandler(): Successfully connected to CastingPlayer (ID: 88665A3712045650) using Commissioner-Generated passcode

### CODE REVIEW

File: examples/tv-casting-app/linux/simple-app-helper.cpp [hunk 1]
```diff
@@ -33,6 +33,8 @@
 
 // VendorId of the Endpoint on the CastingPlayer that the CastingApp desires to interact with after connection
 const uint16_t kDesiredEndpointVendorId = 65521;
+// Indicates that the Commissioner-Generated passcode commissioning flow is in progress.
+bool gCommissionerGeneratedPasscodeFlow = false;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/linux/simple-app-helper.cpp:37]:
```suggestion
bool gCommissionerGeneratedPasscodeFlowRunning = false;
```

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.h [hunk 2]
```diff
@@ -145,12 +145,10 @@ class CastingPlayer : public std::enable_shared_from_this<CastingPlayer>
      * CommissionableDataProvider
      * (matter::casting::core::CastingApp::GetInstance()->UpdateCommissionableDataProvider(CommissionableDataProvider)).
      *
-     * @param connectionCallbacks contains the ConnectCallback and CommissionerDeclarationCallback (Optional).
-     * @param commissioningWindowTimeoutSec (Optional) time (in sec) to keep the commissioning window open, if commissioning is
-     * required. Needs to be >= kCommissioningWindowTimeoutSec.
+     * The same connectionCallbacks and commissioningWindowTimeoutSec parameters passed in to VerifyOrEstablishConnection() will be
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/tv-casting-common/core/CastingPlayer.h:148]: In the docs above for VerifyOrEstablishConnection, can you add a mention of how ContinueConnecting is related to the former i.e. in what cases / on what callbacks may ContinueConnecting need to be called?

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp [hunk 3]
```diff
@@ -272,6 +258,8 @@ CHIP_ERROR CastingPlayer::SendUserDirectedCommissioningRequest()
 
     chip::Protocols::UserDirectedCommissioning::IdentificationDeclaration id = mIdOptions.buildIdentificationDeclarationMessage();
 
+    ReturnErrorOnFailure(support::ChipDeviceEventHandler::SetUdcStatus(true));
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:261]: Why is this needed?

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp [hunk 4]
```diff
@@ -178,33 +178,19 @@ void CastingPlayer::VerifyOrEstablishConnection(ConnectionCallbacks connectionCa
...
-    else
-    {
-        ChipLogProgress(
-            AppServer,
-            "CastingPlayer::VerifyOrEstablishConnection() CommissionerDeclarationCallback not provided in ConnectionCallbacks");
+        ChipLogProgress(AppServer,
+                        "CastingPlayer::ContinueConnecting() CommissionerDeclaration message callback has not been set.");
     }
+    mConnectionState     = CASTING_PLAYER_CONNECTING;
+    mTargetCastingPlayer = this;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:193]: Since we are copying data members over in multiple places now, can you add a operator=() implementation in CastingPlayer that does the copy of the data members as necessary so that we don't skip copying something over mistakenly in one place or another?

File: examples/tv-casting-app/linux/simple-app-helper.cpp [hunk 5]
```diff
@@ -244,6 +246,22 @@ CHIP_ERROR InitCommissionableDataProvider(LinuxCommissionableDataProvider & prov
...
+    ChipLogProgress(AppServer, "simple-app-helper.cpp::PrintEndpointsVendorAndProductIDs() Number of Endpoints: %d",
+                    static_cast<int>(endpoints.size()));
+    for (const auto & endpoint : endpoints)
+    {
+        int endpointId = endpoint->GetId();
+        int vendorId   = endpoint->GetVendorId();
+        int productId  = endpoint->GetProductId();
+        ChipLogProgress(AppServer,
+                        "simple-app-helper.cpp::PrintEndpointsVendorAndProductIDs() Endpoint ID: %d, Vendor ID: %d, Product ID: %d",
+                        endpointId, vendorId, productId);
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/linux/simple-app-helper.cpp:261]: Add a LogDetail() function in the Endpoint class that you can just call from here.

File: examples/tv-casting-app/linux/simple-app-helper.cpp [hunk 6]
```diff
@@ -256,23 +274,49 @@ void ConnectionHandler(CHIP_ERROR err, matter::casting::core::CastingPlayer * ca
...
+
     // Find the desired Endpoint and auto-trigger some Matter Casting demo interactions
     auto it = std::find_if(endpoints.begin(), endpoints.end(),
                            [](const matter::casting::memory::Strong<matter::casting::core::Endpoint> & endpoint) {
-                               return endpoint->GetVendorId() == kDesiredEndpointVendorId;
+                               if (gCommissionerGeneratedPasscodeFlow)
+                               {
+                                   // Since the tv-app with Vendor ID 1111 does not implement the endpoint with Vendor ID 65521 we
+                                   // will (temporarily) use the first available endpoint for demo interactions. See
+                                   // connectedhomeip/examples/tv-app/tv-common/include/AppTv.h.
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/linux/simple-app-helper.cpp:305]:
```suggestion
                                   // Since the tv-app with Vendor ID 1111 does not implement the endpoint with Vendor ID 65521 we
                                   // will (temporarily) use the endpoint with ID 1 for demo interactions. See
                                   // connectedhomeip/examples/tv-app/tv-common/include/AppTv.h.
```

File: examples/tv-casting-app/linux/simple-app-helper.cpp [hunk 7]
```diff
@@ -364,6 +413,13 @@ CHIP_ERROR CommandHandler(int argc, char ** argv)
                                     "Commissioner-Generated Passcode commissioning flow",
                                     index);
                     idOptions.mCommissionerPasscode = true;
+
+                    // Override the default target Vendor ID for this example (Commissioner-Generated passcode commissioning flow).
+                    // This tv-app target Vendor ID (1111), does not implement the AccountLogin cluster, which would otherwise auto
+                    // commission using the Commissionee-Generated passcode upon recieving the IdentificationDeclaration Message.
+                    // See connectedhomeip/examples/tv-app/tv-common/include/AppTv.h.
+                    targetAppInfo.vendorId             = 1111;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/linux/simple-app-helper.cpp:421]: Since this endpoint is not reporting its vendorId, why are we setting this targetAppInfo.vendorId?

File: examples/tv-casting-app/linux/simple-app-helper.cpp [hunk 8]
```diff
@@ -244,6 +246,22 @@ CHIP_ERROR InitCommissionableDataProvider(LinuxCommissionableDataProvider & prov
                          options.payload.discriminator.GetLongValue());
 }
 
+void PrintEndpointsVendorAndProductIDs(
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/linux/simple-app-helper.cpp:249]: I think you could rename to something more succinct like the following:

```suggestion
void LogEndpoints(
```

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp [hunk 9]
```diff
@@ -29,9 +29,7 @@ namespace core {
 
 CastingPlayer * CastingPlayer::mTargetCastingPlayer = nullptr;
 
-// TODO: Verify why commissioningWindowTimeoutSec is a "unsigned long long int" type. Seems too big.
-void CastingPlayer::VerifyOrEstablishConnection(ConnectionCallbacks connectionCallbacks,
-                                                unsigned long long int commissioningWindowTimeoutSec,
+void CastingPlayer::VerifyOrEstablishConnection(ConnectionCallbacks connectionCallbacks, uint64_t commissioningWindowTimeoutSec,
```
Comment (andy31415) [inline examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:32]: comment was removed ... so why are u32 seconds not enough (2^32 seconds is a bit over 136 years ...)

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.h (hunk 2 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/tv-casting-common/core/CastingPlayer.h:148]: Yes.

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp (hunk 9 above)
Comment (andy31415) [reply examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:32]: For prior art, in the chip specification, the `ComissioningTimeout` for `OpenCommissioningWindow` is a u16 (i.e. up to about 18 hours)

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp (hunk 3 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:261]: In my previous PR I moved this from here to VerifyOrEstablishConnection and ContinueConnecting but it's more efficient to have it here once. It previously wasn't called when VerifyOrEstablishConnection was called with CommissionerPasscode set to true for the Commissioner-Generated passcode commissioning flow.
https://github.com/project-chip/connectedhomeip/pull/33479/files#diff-a4c1c744f46eb83abef9f14adb8b60b2e53d0dc8e0d4b846fab4860fbe36b4dfL174

File: examples/tv-casting-app/linux/simple-app-helper.cpp (hunk 5 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/linux/simple-app-helper.cpp:261]: sure.

File: examples/tv-casting-app/linux/simple-app-helper.cpp (hunk 8 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/linux/simple-app-helper.cpp:249]: sure

File: examples/tv-casting-app/linux/simple-app-helper.cpp (hunk 7 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/linux/simple-app-helper.cpp:421]: Even though the Linux tv-app is not reporting this vendorId we know it is valid since it is defined in connectedhomeip/examples/tv-app/tv-common/include/AppTv.h:

    ContentAppImpl mContentApps[APP_LIBRARY_SIZE] = {
        ContentAppImpl("Vendor1", 1, "exampleid", 11, "Version1", "34567890"),
        ContentAppImpl("Vendor2", 65521, "exampleString", 32768, "Version2", "20202021"),
        ContentAppImpl("Vendor3", 9050, "App3", 22, "Version3", "20202021"),
        ContentAppImpl("TestSuiteVendor", 1111, "applicationId", 22, "v2", "20202021")
    };

We need to use this endpoint to demo the Commissioner-Generated passcode flow since this endpoint doesn't have AccountLogin implemented. The default target tv-app with vendorId 65521 would otherwise auto-commission using AccountLogin and the hardcoded passcode "20202021".

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp (hunk 4 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:193]: sure.

File: examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp (hunk 9 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:32]: Yes, I confirmed with sharadb-amazon Sharad Binjola that uint16_t/18 hours is enough for a ComissioningTimeout. Not sure why we had such a big type before. Thanks!

**Review by sharadb-amazon — APPROVED** [review]

File: examples/tv-casting-app/linux/simple-app-helper.cpp [hunk 10]
```diff
@@ -256,23 +268,49 @@ void ConnectionHandler(CHIP_ERROR err, matter::casting::core::CastingPlayer * ca
...
     // Find the desired Endpoint and auto-trigger some Matter Casting demo interactions
     auto it = std::find_if(endpoints.begin(), endpoints.end(),
                            [](const matter::casting::memory::Strong<matter::casting::core::Endpoint> & endpoint) {
-                               return endpoint->GetVendorId() == kDesiredEndpointVendorId;
+                               if (gCommissionerGeneratedPasscodeFlowRunning)
+                               {
+                                   // Since the tv-app with Vendor ID 1111 does not implement the endpoint with Vendor ID 65521 we
+                                   // will (temporarily) use the endpoint with ID == 1 for demo interactions. See
+                                   // connectedhomeip/examples/tv-app/tv-common/include/AppTv.h.
+                                   return endpoint->GetId() == 1;
```
Comment (andy31415) [inline examples/tv-casting-app/linux/simple-app-helper.cpp:300]: add a constant for 1 since we have a cohnstant for the other value?

The comment says that tv app with vendor id 1111, however I don't see that check here. Should we check for GetVendorId == 1111? Generally asking why the comment about VendorID is here ... also the comment is a bit unclear to me: what is the tv app vendorid and what is the endpoint vendorid? Which one is 1111 and which one is 65521?

**Review by andy31415 — APPROVED** [review]
Approving as general readability - I am not familiar with the implementation details of the casting code.

File: examples/tv-casting-app/linux/simple-app-helper.cpp (hunk 10 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/linux/simple-app-helper.cpp:300]: Add in a constant in the next rev.

To demonstrate the Commissioner-Generated passcode commissioning flow we needed to connect to the Target Content Application with Vendor ID 1111, which doesn't implement the AccountLogin cluster (account login obtains the commissionee passcode form a backend server).

The example tv-app simulates 4 content apps installed on a TV device (see https://github.com/pgregorr-amazon/connectedhomeip/blob/master/examples/tv-app/tv-common/include/AppTv.h#L143-L147). Each content app has a Vendor ID defined (1, 65521, 9050 and 1111). After commissioning with a specific content app, the content app reports the Endpoints available for interactions. Each Endpoint has a Endpoint ID and Endpoint Vendor ID attributes.
